### PR TITLE
[AIRFLOW-2425] Maintain lineage and make available in other systems

### PR DESCRIPTION
**Pull request by bolkedebruin**
Make sure you have checked _all_ steps below.

### JIRA
- [ ] My PR addresses the following [Airflow JIRA](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
    - https://issues.apache.org/jira/browse/AIRFLOW-2425
    - In case you are fixing a typo in the documentation you can prepend your commit with \[AIRFLOW-XXX\], code changes always need a JIRA issue.

### Description
- [X] Here are some details about my PR, including screenshots of any UI changes:

Lineage 

### Tests
- [X] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits
- [X] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

### Documentation
- X ] In case of new functionality, my PR adds documentation that describes how to use it.
    - When adding new operators/hooks/sensors, the autoclass documentation generation needs to be added.

### Code Quality
- [X] Passes `git diff upstream/master -u -- "*.py" | flake8 --diff`

Will squash when tests pass and comments are integrated.

cc: @criccomini @jgao54 @Fokko 

screen shot 2018-05-05 at 23 54 55

### CODE REVIEW

File: airflow/lineage/datasets.py [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+from jinja2 import Environment
+
+
+def _inherited(cls):
+    return set(cls.__subclasses__()).union(
+        [s for c in cls.__subclasses__() for s in _inherited(c)]
+    )
+
+
+def map_type(name):
```
Comment (Fokko) [inline airflow/lineage/datasets.py]: Move this as a static method inside the `DataSet` class

File: airflow/lineage/datasets.py [hunk 2]
```diff
@@ -0,0 +1,140 @@
+# -*- coding: utf-8 -*-
```
Comment (Fokko) [inline airflow/lineage/datasets.py:1]: Can we split the different files to different files? I maybe move these files in the dataset module.

File: tests/lineage/test_lineage.py [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+
+            op1.set_downstream(op3)
+            op2.set_downstream(op3)
+            op3.set_downstream(op4)
+            op4.set_downstream(op5)
+
+        ctx1 = {"ti": TI(task=op1, execution_date=DEFAULT_DATE)}
+        ctx2 = {"ti": TI(task=op2, execution_date=DEFAULT_DATE)}
+        ctx3 = {"ti": TI(task=op3, execution_date=DEFAULT_DATE)}
+        # ctx4 = {"ti": TI(task=op4, execution_date=DEFAULT_DATE)}
```
Comment (Fokko) [inline tests/lineage/test_lineage.py]: Why is this one commented out?

File: docs/lineage.rst [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+
+Airflow can help track origins of data, what happens to it and where it moves over time. This can aid having
+audit trails and data governance, but also debugging of data flows.
+
+Airflow tracks data by means of inlets and outlets of the tasks. Let's work from an example and see how it
+works.
+
+.. code:: python
+
+    """
```
Comment (Fokko) [inline docs/lineage.rst]: Missing closing quotes of the comments? Maybe move the comments outside of the code block

File: docs/lineage.rst [hunk 5]
```diff
@@ -0,0 +1,87 @@
+Lineage
```
Comment (Fokko) [inline docs/lineage.rst:1]: Maybe mention it is still in alpha?

File: airflow/lineage/backend/atlas/__init__.py [hunk 6]
```diff
@@ -0,0 +1,96 @@
+# -*- coding: utf-8 -*-
```
Comment (Fokko) [inline airflow/lineage/backend/atlas/__init__.py:1]: For me it makes more sense that we have an interface that is extended by an implementation of a lineage backend.

File: airflow/operators/bash_operator.py [hunk 7]
```diff
@@ -89,6 +89,7 @@ def execute(self, context):
         bash_command = ('export {}={}; '.format(AIRFLOW_HOME_VAR, airflow_home_value) +
                         'export {}={}; '.format(PYTHONPATH_VAR, pythonpath_value) +
                         self.bash_command)
+        self.lineage_data = bash_command
```
Comment (Fokko) [inline airflow/operators/bash_operator.py:92]: Can't we also do this using a decorator, like we do with the `{pre,post}_execute`.

**Review by Fokko — CHANGES_REQUESTED** [review]

File: airflow/lineage/datasets.py (hunk 1 above)
Comment (bolkedebruin) [reply airflow/lineage/datasets.py]: will do

File: airflow/lineage/datasets.py (hunk 2 above)
Comment (bolkedebruin) [reply airflow/lineage/datasets.py:1]: I think that is a bit over kill. At the moment datasets are designed to be light weight so it doesnt make much sense to split it out right now.

File: tests/lineage/test_lineage.py (hunk 3 above)
Comment (bolkedebruin) [reply tests/lineage/test_lineage.py]: No test required I can remove it

File: docs/lineage.rst (hunk 5 above)
Comment (bolkedebruin) [reply docs/lineage.rst:1]: It will be in master not in a release yet

File: airflow/lineage/backend/atlas/__init__.py (hunk 6 above)
Comment (bolkedebruin) [reply airflow/lineage/backend/atlas/__init__.py:1]: You mean using an abstract class? It looks like overkill right now (only one function is called), but can do

File: airflow/operators/bash_operator.py (hunk 7 above)
Comment (bolkedebruin) [reply airflow/operators/bash_operator.py:92]: How to do that? It is an arbitrary assignment, so how do I know which variable to catch?

**Comment by bolkedebruin** [comment]
@Fokko PTAL

File: airflow/operators/email_operator.py [hunk 8]
```diff
@@ -35,7 +35,7 @@ class EmailOperator(BaseOperator):
     :type html_content: string
     :param files: file names to attach in email
     :type files: list
-    :param cc: list of recipients to be added in CC field
+    :param cc: list of recipients to be added in CC fieldn
```
Comment (msumit) [inline airflow/operators/email_operator.py]: typo `fieldn`?

Comment (bolkedebruin) [reply airflow/operators/email_operator.py]: Good catch. Should not have changed this file

**Review by Fokko — APPROVED** [review]
LGTM

**Pull request closed without merging**